### ISSUE TITLE
Batch computations

### ISSUE THREAD

**Issue by peastman** [openmm/openmm#2535]
Batch computation is a feature that's been requested a few times.  The idea is that you want to run many simulations of identical systems, or perhaps just evaluate the energies of many conformations.  For a relatively large system you can just loop over them, but for smaller systems the overhead becomes prohibitive.

This is especially significant when using a high end GPU.  A system with just a few hundred atoms is far too small to take advantage of it, so most of the compute units end up sitting idle.  But if you could evaluate the energies of 100 different conformations at the same time, that would make far better use of it.

This raises a lot of issues, both in terms of the architecture and the API.  To minimize disruption to the code, it might be simplest if we effectively build a giant system consisting of 100 copies of the molecule.  Bonded interactions will be straightforward, but nonbonded interactions need to be handled in an efficient way so it doesn't try to compute interactions between different copies.

Understanding the use cases will be critical.  For example, what nonbonded methods does this need to support?  A no cutoff version would be pretty easy to implement, while a PME version would be much more complicated.

**Comment by giadefa** [comment]
one case where batch computation is useful and we use it outside openMM is
neural network potentials. In this case, it is natural to batch the NN
force calculations while running multiple replicas at once. There is
nothing else that local bonds, angles and some repulsion. Maybe in the
future it would be useful to have simple cutoff based Coulomb for now.

g

**Comment by proteneer** [comment]
One approach I've found useful is to just overload the batch dimension along the kernels's blockIdx.y, and just re-index appropriately. This is significantly easier than having to combine them all into a single calculations.

**Comment by peastman** [comment]
A particularly simple API would be to just add a `setNumCopies()` method to System.  The vectors you pass to `setPositions()` and `setVelocities()` would then have length (# particles)x(# atoms), so existing code would continue to work without change.  The accessor methods in State would get an argument for the copy index, with a default value of 0 to maintain source compatibility.

An initial implementation could just loop over the copies and compute their forces and energies individually.  So it wouldn't speed up that computation, but it would still save overhead in setting positions and retrieving forces.  We could then optimize individual forces, focusing on just the ones that are bottlenecks for important use cases.  The TensorFlow and PyTorch plugins would be particularly easy to optimize: just add another dimension to the input and output tensors.

**Comment by jchodera** [comment]
(Migrated from #2822)

As GPUs get wider, it becomes more difficult to achieve high utilization of the entire GPU. New architectures like Ampere could help multiple `Context`s running in separate threads take better advantage of the GPU, but I wonder what features we could add to OpenMM to better enable it to exploit wide GPUs.

**Are there useful features we could build in that would make management of multiple simulations easier?** Right now, an application could manage multiple threads with independent `Context`s on the same GPU. Under which circumstances would this be efficient? Is there anything we can do at the OpenMM API or `app` level to make it easier to manage multiple simulations simultaneously, such as adding convenience methods for managing multiple simulations, or building in enhanced sampling algorithms like replica exchange that would better exploit wide GPUs?

**Are there useful low-level features we could consider that would allow multiple simulations to be packed into the same `Context`?** One could conceive of defining a single `System` object that actually contains `N` concatenated simulation systems, and providing a way to define how blocks of particles should be considered independent, noninteracting systems. Pairlist construction would consider each block separately in computing interactions, and methods like PME would be computed independently for each block. But this could allow a set of `N` small systems to be effectively simulated in parallel.

**Comment by dmclark17** [comment]
I wanted to follow up on the possibility of concurrently executing multiple simulations. Running multiple simulations with separate Python instances and MPS enabled gave a significant boost to the total ns/day for smaller simulations. However, since MPS doesn't support Windows, this approach wouldn't necessarily work for the F@H use-case. 

Is there still interest in pursuing this? If so, I can look into the best way to support it in the CUDA backend.

**Comment by joseph-c-kim** [comment]
@peastman directed me here to discuss my use-case.

Given the same simulation context, I wish to compute potential energies and forces for different conformations of a protein system in parallel. My current setup is to just run `context.setPositions(x)` followed by `context.getState(getForces=True, getEnergy=True)` for different conformations `x` in a for loop. This works, but I was just wondering if there is a more efficient way to do this given that the simulation context does not change at all for each input set of positions x.

For further context, I checked openmm-torch (since I am primarily using OpenMM with PyTorch), but it doesn't seem like they have support for batch computations either.

Another thing I've tried is using python's `multiprocessing` module. While this does provide speed-up when using the Reference platform, it did not provide much of a speedup when I'm using a GPU with the CUDA platform (which is my preferred use-case as I have relatively large systems and have access to multiple GPUs).

It seems that this has been a requested feature since 2 years ago. Are there any plans to support batch computations? If not, is there a workaround I can do currently?

**Comment by peastman** [comment]
A few questions about your simulations:

- How many atoms do they have?
- Are the forces computed entirely by PyTorch, or do you also include some regular forces?
- If the latter, which forces are the most expensive part of your computation?

The main benefit of this feature is for very small systems on a GPU.  In that case, there isn't enough work to keep the GPU busy and the time is dominated by overhead.  Being able to do several computations at once reduces the overhead and leaves less of the GPU idle.  For large systems there's unlikely to be much benefit, since there's already plenty of work to keep the GPU busy.

**Comment by joseph-c-kim** [comment]
I'm looking at proteins as small as ~10 residues and as large as ~150 residues. I am only using the force and energy computed in OpenMM (not from PyTorch) from a system that I create with an Amber prmtop file. No external forces.

**Comment by peastman** [comment]
I must have misunderstood.  I thought you said you were using the openmm-torch plugin to interface with pytorch models?

Since you're simulating smallish proteins on a large GPU, there could be some benefit to computing multiple conformations at once.  A single computation probably won't be enough to saturate the GPU.  You might try using MPS.  @dmclark17 mentioned above that he got a good speedup from it.

**Comment by joseph-c-kim** [comment]
I haven't been using the openmm-torch plugin. I only recently learned about it in my search for batch computation solutions, but it doesn't seem like there's support for such a feature in the plugin. 

I'm not too familiar with MPS. This is for CUDA multiprocessing, correct? Is there a Python library to directly utilize MPS?

**Comment by peastman** [comment]
MPS is a way of splitting up a GPU so multiple processes can share it more effectively.  I haven't used it either so I'm not too familiar with the details.  https://github.com/lebedov/cudamps looks like it might be useful.

**Comment by dmclark17** [comment]
MPS should improve performance when using multiple processes per GPU. I have seen performance benefits when launching multiple Python processes running OpenMM:

```
numactl --membind=0 --physcpubind=0 python benchmark.py --platform=CUDA --test=gbsa --seconds=30 &
numactl --membind=0 --physcpubind=1 python benchmark.py --platform=CUDA --test=gbsa --seconds=30 &
wait
```

MPS can be started using:
```
export CUDA_MPS_PIPE_DIRECTORY=/tmp/nvidia-mps
export CUDA_MPS_LOG_DIRECTORY=/tmp/nvidia-log
CUDA_VISIBLE_DEVICES=0 nvidia-cuda-mps-control -d
sleep 5 # Wait for MPS to start
```

And exited with:
```
echo quit | nvidia-cuda-mps-control
```

It looks like the library Peter linked is using these commands, and more information can be found here: https://docs.nvidia.com/deploy/mps/index.html#topic_5_1_1

**Comment by joseph-c-kim** [comment]
Ah okay. I'm not using OpenMM to run simulations though. I'm using it in a deep learning setting, where I would like to compute energies and forces of conformations generated at iterations.

**Comment by peastman** [comment]
You could try using cudamps or something similar to send off multiple conformations to different processes and compute them in parallel.  You said you didn't see much benefit from doing that with `multiprocessing`, but it's possible MPS would make it more efficient.

Then again, it might not.  It depends what the bottlenecks are, and how much overhead is added by using multiple processes.

Another thing you could try is creating multiple Contexts in a single process and evaluating them in parallel on multiple threads.  `getState()` releases the global interpreter lock, so multiple threads can be executing at once.  That will have less overhead than using multiple processes.  Just be careful to do it in a thread safe way.  A single Context should never be accessed from two threads at the same time.

**Comment by philipturner** [comment]
I was directed here to discuss my use case (https://github.com/openmm/openmm/issues/3942)

I'm evaluating whether I can use OpenMM for a genetic algorithm. Simulate many small simulations, each possibly with a different number of atoms. I don't have an NVIDIA GPU, so CUDA MPS won't help. Even with MPS, encoding each simulation separately utilizes several CPU cores. Command encoding would become a bottleneck (worse the smaller a simulation gets). Running multiple independent `Context`s simultaneously could not scale to under ~1000 atoms.

Perhaps we make a new OpenMM public API, called `GroupContext` for closely related replica simulations. The user chooses whether each has the same number of atoms and/or same bounding box, an optimization but not a requirement. This would require major changes to the common compute kernels: compile macros for whether you're in multi-sim mode. If the `HIP` and `Metal` plugins don't incorporate such macros, it's simply an unsupported feature like some forcefields. We also assume each simulation has similar computational cost.

Bonus: We could make another feature, where the simulations share most computations, but sometimes have slightly different forcefields.

**Comment by axsk** [comment]
I also have a great interest in running many copies of short random simulations (starting from different initial conditions) simultaneously.

I tried to go with @peastman's suggestion 
> Another thing you could try is creating multiple Contexts in a single process and evaluating them in parallel on multiple threads. `getState()` releases the global interpreter lock, so multiple threads can be executing at once. That will have less overhead than using multiple processes. Just be careful to do it in a thread safe way. A single Context should never be accessed from two threads at the same time.

But I don't know how to create multiple copies of the same system properly.
`copy.deepcopy` is not supported and `copy.copy(simulation)` is not enough as the simulations quickly reach NaN positions (and "infect" each other immediately) (maybe also related to #4393?)
Here is what i tried: [code](https://github.com/axsk/ISOKANN.jl/blob/0ba14c6808e07c0c1c15745d3fb1226a501753be/src/openmm.jl#L48)

Any suggestions on how to construct the different "worker simulations"?

**Comment by philipturner** [comment]
Here's a great approach. Provided you aren't using PME, and the nonbonded forces are purely local. Make a massive system with many smaller systems, each separated farther than the nonbonded cutoff. Then they don't interact, making it like several isolated simulations. 

I've used the approach to energy-minimize several dozen nano-parts in a single simulation. Each part is ~5,000 atoms, yet the latency to hear from GPU is so large that 5,000 atoms doesn't saturate the GPU. One energy minimization often takes ~1-2 seconds. Running several dozen minimizations in parallel helps to reduce the latency.

**Comment by RaulPPelaez** [comment]
The problem with this is that depending on system size and copies you will get into numerical accuracy issues when the origin of a copy is too far away from 0.

**Comment by philipturner** [comment]
It depends on which application you're using. If the truncation of the mantissa is a problem, this might not work. You have to analyze exactly what parts of the code contribute most to the error, and rank them in order of importance. I imagine this is not a big contributor.

**Comment by peastman** [comment]
As long as you aren't modifying the System, it's fine to share it between threads.  It has no mutable fields that change during a simulation.  Each thread should create its own Integrator and Simulation, and they should only be accessed from that thread.

**Comment by axsk** [comment]
Creating individual `Context()` objects with `copy.copy(integrator)` and a shared `system` (as well as appropriate position and velocities) worked out fine :)
I am not sure how this will interact with GPU yet.

**Comment by peastman** [comment]
It should work ok with the GPU, at least for a limited number of threads.  GPUs aren't as good at multitasking as CPUs, though they're better at it than they used to be.  You may also find there's a hard limit on how many contexts you can create before the GPU runs out of resources to create new ones.

**Comment by giadefa** [comment]
it would be nice to have some benchmarks

**Comment by philipturner** [comment]
Apple is the primary victim of this latency bottleneck. Their drivers have very high latencies, and little opportunity for running multiple GPU commands in parallel. AMD is in the middle ground. Meanwhile, Nvidia has extremely low (10 µs) latencies so even the tiniest problem can fully harness a GPU. For the benchmarked devices ([RTX 4090, A100, etc.](https://openmm.org/benchmarks)) the results will be skewed toward "there is no need for batching".

**Comment by giadefa** [comment]
I have heard the opposite for 4090 and alike, so it would be nice to have
numbers

**Comment by peastman** [comment]
RTX 4090 is a very wide GPU.  It takes quite a large system to saturate it.  For small systems, being able to run multiple simulations in parallel should definitely help.

**Comment by philipturner** [comment]
Nvidia has multi-instance GPU but none of the other vendors have that.

**Comment by pipitoludovico** [comment]
Hi everyone!
I am playing around a bit with Python's multiprocessing and multiple GPUs on the same thread. 
I will try the approach used by @axsk and see how it will pan out with multiple GPUs.

So far, the multi-GPU thread/simulation will fail after some time (>100 ns) resulting in NaN and calling for the error handlers (gsd's __init__ and multiprocessing's connection). Specifically, this issue happens only when running with more than 1 Nvidia GPU (I have tried with 2,3,4 devices at the same time in various combinations to exclude some specific hardware fault).

If the context solution works, I'll close issue #4393 with that approach, although this is something that might still need a closer look.

**Comment by axsk** [comment]
I need really short trajectories, so that the overhead for Context() creation became a problem. I now use checkpoints to improve that performance.
IIRC using multiple threads via Joblib did not bring the expected performance gains.

And I think I also ran into NaN issues( 1 GPU, multithreaded OpenMM calls) but have yet to confirm this is due to multithreading and/or GPU.

For reference [here](https://github.com/axsk/ISOKANN.jl/blob/4f8399a9f6c5cadb095733d8a55f3d01ccc025e4/src/openmm.jl#L90) is how I run this.

**Comment by peastman** [comment]
> c = copy.copy(sim.context)

That isn't going to work correctly.  Contexts are not copyable.  You can copy the Integrator, but then you need to create a new Context properly:

```python
c = Context(context.getSytem(), copy.copy(context.getIntegrator()), context.getPlatform())
```

You don't specify the platform, which means it chooses one automatically.  Which is it using?  How large are your systems?  CUDA and OpenCL take the most time to create a Context.  CPU takes much less time, and Reference takes even less.  If the systems are small, one of them might end up being faster overall.

**Comment by CheukHinHoJerry** [comment]
Hi everyone, I am trying to run batch computations too but I encounter a problem. Following @axsk 's suggestions, I am doing something like this:

```python
from openmm.app import *
from openmm import *
from openmm.unit import *
from multiprocessing.context import Process
import openmmtools
import multiprocessing

def run_steps(integrator, NumSteps):
    print(integrator)
    integrator.step(NumSteps)

if __name__ == "__main__":
    multiprocessing.set_start_method('spawn')
    temp = 300
    NumIntegrators = 1
    pdb = PDBFile('data/ala3_vac_capped.pdb')
    forcefield = ForceField('amber99sb.xml')#, 'tip3p.xml')
    TotalTime = 0.1*picosecond# in ps
    dt = 1.0 * femtosecond

    # fine system, position and topology # NoCutoff
    system = forcefield.createSystem(pdb.topology, nonbondedMethod=CutoffNonPeriodic, nonbondedCutoff=1*nanometer, constraints=HBonds)
    system, positions, topology = system, pdb.positions, pdb.topology
    npos = len(pdb.positions)
    NumSteps = 5

    ##

    # Define integrators
    integrators = [openmmtools.integrators.LangevinIntegrator(temp * kelvin, 1 / picosecond, dt) for _ in range(NumIntegrators)]
    contexts = [Context(system, fi) for fi in integrators]

    print("Done context assmble")
    for cc in contexts:
        cc.setPositions(positions)
        
    print("Done setting positions")
    print(integrators[0])
    
    processes = []
    for (n, _integrator) in enumerate(integrators):
        process = Process(target = run_steps, args=(_integrator, NumSteps))
        processes.append(process)
        process.start()

    for p in processes:
        p.join()
```

But it seems that everything I create a `Process`, a new `integrator` is created and the new integrator is not bounded to a context, which causes error. How can I get around with this?

**Comment by philipturner** [comment]
Python is not multithreaded. You’ll need to use either C++ or Swift to attempt that approach.

**Comment by axsk** [comment]
@peastman You were right, I had to create new Contexts.

@philipturner I believe multithreading works fine (using Joblib out of familiarity). 
The basic idea is that the heavy lifting is happening in the (non-blocking) calls to the (binary) OpenMM library, such that the Python runtime is free to start multiple threads each calling those.

I observed considerable speedup using multiple threads.

**Comment by axsk** [comment]
> But it seems that everything I create a `Process`, a new `integrator` is created and the new integrator is not bounded to a context, which causes error. How can I get around with this?

I suppose you should create both, the integrator and the context, in each individiual process anew (e.g. in your `run_steps` function, see also `singlerun` in my code), so they have consistent references.

**Comment by CheukHinHoJerry** [comment]
> > But it seems that everything I create a `Process`, a new `integrator` is created and the new integrator is not bounded to a context, which causes error. How can I get around with this?
> 
> I suppose you should create both, the integrator and the context, in each individiual process anew (e.g. in your `run_steps` function, see also `singlerun` in my code), so they have consistent references.

Thank you - I think what you suggest will work fine for running long simulations. But in my use case that has a problem because I am implementing a new algorithm that has to call small number of steps in parallel and then communicate between workers many times. And creating `Context` take a non-trivial amount of time.

**Comment by peastman** [comment]
Python provides two different mechanisms for parallelism: processes and threads.  You're using the `multiprocessing` module, which creates processes.  Processes do not share memory with each other.  Every object is only visible to the process that created it.  If you pass an object between processes, really it means that the original object gets serialized and a new object is created from the serialized representation on the other side.

Threads are a lighter weight mechanism.  They exist within a single process.  They share memory with each other, so multiple threads can access the same object.  That can be much more efficient, although it also means you need to be careful about two threads modifying the same object at the same time.

Threads have another big disadvantage that's specific to Python: only one thread can execute Python code at a time.  There's something called the Global Interpreter Lock (GIL) that can only be held by one thread at a time, and only the thread holding it executes.  That makes threading a lot less useful than in other languages, and it's the main reason people often use processes instead of threads.

But most of the expensive parts of an OpenMM simulation are written in C++, and they aren't bound by that limitation.  When you call `step()` on an Integrator, it releases the GIL.  That lets another Python thread start running.  If that thread calls `step()` on another Integrator, it too releases the GIL and a third thread can start running.

For what you're trying to do, threads are probably a better choice.

This is fairly far removed from the subject of this issue.  If you want to discuss it further, can you open a new issue?

**Comment by arisu3** [comment]
For CUDA at least, would it be possible to use multiple Streams to accomplish this, similar to the way that the PME calculations can be done in its own stream?

**Comment by peastman** [comment]
Streams can be effective for simulating just a few copies.  For large numbers of copies, you really want to calculate all of them in the same kernel.

**Comment by arisu3** [comment]
How many is "a few" roughly? I assume the limit would be similar for just running multiple instances of OpenMM using MPS, i.e. running several copies of small systems to improve occupancy and overall GPU utilization would be fine with streams, but running, say, thousands of extremely tiny systems is something that should be done in one kernel?

**Comment by peastman** [comment]
That's correct.  Streams and MPS are both ways of getting the GPU to execute multiple kernels at once.  They have similar limits in how much parallelism they can get.

**Comment by arisu3** [comment]
What is the feasibility of allowing a single OpenMM process to run multiple systems in separate streams? Although the parallelism isn't perfect, it would be of great help to FAH where even running three or four simultaneous systems would massively increase the utilization of wider GPUs.

A big reason why MPS isn't being considered as the standard way to do it is that it is not supported by Windows.

**Comment by peastman** [comment]
OpenMM has no problem creating multiple Contexts on the same device.  It would only require changes to the FAH core.

**Comment by arisu3** [comment]
Are those contexts run concurrently as separate streams?

**Comment by peastman** [comment]
Yes.  Each one has its own `CUcontext` with its own default stream.
